addons/Dexie.Syncable/src/apply-changes.js

```diff
@@ -4,31 +4,35 @@ import bulkUpdate from './bulk-update';
 
 export default function initApplyChanges(db) {
   return function applyChanges(changes, offset) {
-    const length = changes.length;
-    // This is the base case for the recursion
-    if (offset >= length) return Dexie.Promise.resolve(null);
-    const firstChange = changes[offset];
-    let i, change;
-    for (i=offset + 1; i < length; ++i) {
-      change = changes[i];
-      if (change.type !== firstChange.type ||
-          change.table !== firstChange.table)
-        break;
-    }
-    const table = db.table(firstChange.table);
-    const specifyKeys = !table.schema.primKey.keyPath;
-    const changesToApply = changes.slice(offset, i);
-    const changeType = firstChange.type;
-    const bulkPromise =
-        changeType === CREATE ?
-            table.bulkPut(changesToApply.map(c => c.obj), specifyKeys ?
-                changesToApply.map(c => c.key) : undefined) :
-            changeType === UPDATE ?
-                bulkUpdate(table, changesToApply) :
-                changeType === DELETE ?
-                    table.bulkDelete(changesToApply.map(c => c.key)) :
-                    Dexie.Promise.resolve(null);
+    if (offset === changes.length)
+      return Dexie.Promise.resolve(null);
+    changes = changes.slice(offset);
 
-    return bulkPromise.then(()=>applyChanges(changes, i));
+    let collectedChanges = {};
+    changes.map((change) => {
+      if (!collectedChanges.hasOwnProperty(change.table)) {
+        collectedChanges[change.table] = [[], [], [], []];
+      }
+      collectedChanges[change.table][change.type].push(change);
+    });
+    let table_names = Object.keys(collectedChanges);
+    let tables = table_names.map((table) => db.table(table));
+
+    return db.transaction("rw", tables, () => {
```

Comment:
I put a transaction here as it looked like finally-commit-all-changes depended on either all of the operations happen or they don't.